File: 1383-maximum-performance-of-a-team/1383-maximum-performance-of-a-team.js
/**
 * @param {number} n
 * @param {number[]} speed
 * @param {number[]} efficiency
 * @param {number} k
 * @return {number}
 */
var maxPerformance = function(n, speed, efficiency, k) {
    const rank = efficiency.map((e, i) => i).sort((a, b) => efficiency[b] - efficiency[a]);
    
    let heap = new MinPriorityQueue();
    let speedSum = 0n;
    let result = 0n;
    
    for (let i = 0; i < n; i++) {
        heap.enqueue(speed[rank[i]]);
        speedSum = speedSum + BigInt(speed[rank[i]]);
        if (heap.size() > k) {
            speedSum = speedSum - BigInt(heap.dequeue().element);
        }
        let perf = speedSum * BigInt(efficiency[rank[i]])
        if (result < perf) result = perf;
    }
    
    return result % 1000000007n ;
    
};

File: 1383-maximum-performance-of-a-team/1383-maximum-performance-of-a-team.py
from queue import PriorityQueue


class Solution:
    def maxPerformance(self, n: int, speed: List[int], efficiency: List[int], k: int) -> int:
        result = 0
        speedSum = 0
        
        heap = PriorityQueue()
        
        for eff, spd in sorted(zip(efficiency, speed), reverse=True):
            speedSum += spd
            heap.put(spd)
            if heap.qsize() > k:
                speedSum -= heap.get()
            
            result = max(result, speedSum * eff)
        
        
        return result % 1000000007
        
        